File: javascript/noisy_animal.js
function createNoisyAnimal(species) {
  function isBird() {
    return ['owl', 'eagle', 'hadedah'].includes(species);
  }

  function isMammal() {
    return ['cat', 'dog', 'leopard'].includes(species);
  }

  function mammalNoise() {
    const noises = {
      cat: 'meow',
      dog: 'woof',
      leopard: 'growl'
    };
    return noises[species];
  }

  function makeMammalNoise() {
    console.log(mammalNoise());
  }

  function makeBirdNoiseLoud() {
    if (species === 'hadedah') {
      console.log('squawk');
      console.log('squawk');
    } else if (species === 'eagle') {
      console.log('caw');
      console.log('caw');
    } else if (species === 'owl') {
      console.log('hoot');
      console.log('hoot');
    }
  }

  function makeBirdNoiseSoft() {
    if (species === 'hadedah') {
      console.log('squawk');
      throw new Error('there is no such thing as a quiet hadedah!');
    } else if (species === 'eagle') {
      console.log('caw');
    } else {
      console.log('hoot');
    }
  }

  function makeNoise({ loud = true } = {}) {
    if (loud) {
      if (isMammal()) {
        makeMammalNoise();
        makeMammalNoise();
      }
      if (isBird()) {
        makeBirdNoiseLoud();
      }
    } else {
      if (isMammal()) {
        makeMammalNoise();
      } else if (isBird()) {
        makeBirdNoiseSoft();
      }
    }
  }

  return {
    makeNoise
  };
}

module.exports = createNoisyAnimal;

File: ruby/noisy_animal.rb
# frozen_string_literal: true

class NoisyAnimal
  attr_reader :species

  def initialize(species)
    @species = species
  end

  def make_noise(loud: true)
    if species == 'hadedah' && loud
      2.times { puts bird_noise }
    elsif species == 'hadedah' && !loud
      puts bird_noise
      raise 'there is no such thing as a quiet hadedah!'
    elsif is_animal && loud
      2.times { puts animal_noise }
    elsif is_animal && !loud
      puts animal_noise
    elsif is_bird && loud
      2.times { puts bird_noise }
    elsif is_bird && !loud
      puts bird_noise
    end
  end

  private

  def animal_noise
    {
      'cat' => 'meow',
      'dog' => 'woof',
      'leopard' => 'growl'
    }[species]
  end

  def bird_noise
    {
      'hadedah' => 'squawk',
      'eagle' => 'caw',
      'owl' => 'hoot'
    }[species]
  end

  def is_bird
    species == 'owl' || species == 'eagle' || species == 'hadedah'
  end

  def is_animal
    %w[cat dog leopard].include?(species)
  end
end
